README.md
# historical-bank-ruby

[![Gem Version](https://badge.fury.io/rb/historical-bank.svg)](https://badge.fury.io/rb/historical-bank)
[![Build Status](https://travis-ci.org/Skyscanner/historical-bank-ruby.svg?branch=master)](https://travis-ci.org/Skyscanner/historical-bank-ruby)

## Description

This gem provides a bank that can serve historical rates,
contrary to most bank implementations that provide only current market rates.
[Open Exchange Rates](https://openexchangerates.org/) (OER) is used as the provider of the rates.
As the HTTP requests to OER can add latency to your calls, a `RatesStore` (cache) based on Redis was added, making it super-fast.

You can use it as your default bank and keep calling the standard `money` gem methods (`Money#exchange_to`, `Bank#exchange_with`). On top of that, we've added a few more methods that allow accessing historical rates (`Money#exchange_to_historical`, `Bank#exchange_with_historical`).


### Base currency

An **exchange rate** has a **base currency** and a **quote (or counter) currency**.
More specifically, it is the price of 1 unit of base currency in the quote currency.
For example, if base currency is EUR, quote currency is USD, and the rate is 1.25,
this means that 1 EUR is equal to 1.25 USD.

All the rates fetched and cached by this `Bank` are relative to a single base currency which is defined in the configuration block.
This helps to optimize fetching and caching rates.

Default base currency is EUR, but it can be changed in the config.


### Dates, times, timezones

The timezone used throughout this gem is the UTC timezone.

All processed rates (fetched from OER, added manually, cached in Redis) are considered to be the [closing (end of day) rates](https://openexchangerates.org/faq/#eod-values) for their associated dates in UTC.
For example, when we have a cached rate of EUR->USD on January 10th 2017 with value 1.25,
this means that 1 EUR was equivalent to 1.25 USD on January 10th at 23:59:59.
This is the historical rate that the bank will use for exchanging EUR with USD on that date.
Consequently, a rate for a certain `date` fetched from OER becomes available at 00:00 UTC on `date+1`.

For convenience, methods that accept `Date`s as arguments can accept `Time`s as well.
When a `Time` is used, it is first converted into the UTC-equivalent `Date`,
and method is executed as if that `Date` was passed instead.
For example, when the `Time` `2017-01-10 02:50:00 +04:00` is passed as argument to `#exchange_with_historical`,

```ruby
from_money = Money.new(100_00, 'EUR')
to_currency = 'USD'

# 2017-01-10 02:50:00 +0400
bank.exchange_with_historical(from_money, to_currency, Time.new(2017, 1, 10, 2, 50, 0, '+04:00'))
# => #<Money fractional:10585 currency:USD>
```

it is equivalent to passing the `Date` `2017-01-09`

```ruby
# 2017-01-09
bank.exchange_with_historical(from_money, to_currency, Date.new(2017, 1, 9))
# => #<Money fractional:10585 currency:USD>
```


### Caching

We've implemented 2 layers of caching in order to obliterate latency!
First layer is memory (instance variable in the bank object), and second is Redis.
If desired rate is not found in memory, the bank tries to look it up in Redis.
If that fails too, a request to OER is made.

When we fetch rates from OER, they are cached in Redis and memory too.
Similarly, when the rate is found in Redis, it is again cached in memory.

<p align="center">
  <img src="images/cache_diagram.png" alt="Caching diagram"/>
</p>

Pretty simple and fast!


### Singleton

The bank follows the Singleton pattern, as it inherits from `money` gem's `Money::Bank::Base`.
This also helps preserve the memory cache across calls.
Don't worry, the bank is thread-safe!


## Requirements

- **OpenExchangeRates account** - If on the Free or Developer plan, the `/historical/*.json` endpoint is used for fetching rates for single days. If on the Enterprise or Unlimited plan, the `/time-series.json` endpoint is used for fetching rates for whole months, increasing efficiency.
- **Redis >= 2.8** (versions of Redis earlier than 2.8 may also work fine, but this gem has only been tested with 2.8 and above.)
- **Ruby >= 2.0**



## Installation

```
gem install historical-bank
```

Alternatively, if you're using `bundler`, you can add
``` ruby
gem 'historical-bank'
```
to your `Gemfile` and run `bundle install`



## Usage

Example scripts demonstrating all functionality can be found in [`examples/`](examples/).

### Configuration

```ruby
Money::Bank::Historical.configure do |config|
  # (required) your OpenExchangeRates App ID
  config.oer_app_id = 'XXXXXXXXXXXXXXX'

  # (optional) account type on openexchangerate.org (default: 'Enterprise')
  # replace 'FREE' with 'DEVELOPER', 'ENTERPRISE', or 'UNLIMITED', according to your account type.
  config.oer_account_type = Money::RatesProvider::OpenExchangeRates::AccountType::FREE

  # (optional) currency relative to which all the rates are stored (default: EUR)
  config.base_currency = Money::Currency.new('USD')

  # (optional) the URL of the Redis server (default: 'redis://localhost:6379')
  config.redis_url = 'redis://localhost:6379'

  # (optional) Redis namespace to prefix all keys (default: 'currency')
  config.redis_namespace = 'currency_historical_gem'

  # (optional) set a timeout for the OER calls (default: 15 seconds)
  config.timeout = 20
end
```

#### Rails

In Rails, config should be set inside an initializer (`config/initializers`).
If you have the `money-rails` gem installed, you can add this on top of the `config/initializers/money.rb` file.

```ruby
# config/initializers/money.rb

require 'money/bank/historical'

Money::Bank::Historical.configure do |config|
  # ....
end


MoneyRails.configure do |config|
  # ...

  # if you want to set it as the default bank
  config.default_bank = Money::Bank::Historical.instance

  # ...
end
```

#### Sinatra

In a Sinatra app (or any other kind of app), simply set the config before you call `Money::Bank::Historical.instance`.

```ruby
# app.rb

require 'money/bank/historical'

Money::Bank::Historical.configure do |config|
  # ....
end

class App < Sinatra::Base

  configure do
    # if you want to set it as the default bank
    Money.default_bank = Money::Bank::Historical.instance
  end

  # ...
end
```

### Dates

The minimum date for which we can fetch rates is January 1st 1999.
This [limitation](https://docs.openexchangerates.org/docs/api-introduction) is set by the OpenExchangeRates API.
The maximum date for which we fetch OER rates is yesterday (in UTC),
as [today's rates are not yet final](https://openexchangerates.org/faq/#timezone).

However, if you want to overcome these limitations manually, you can add past (and even future!) rates using `Historical#add_rate` and `#add_rates`.

### Exchange

You can use the bank object for the exchange
```ruby
from_money = Money.new(100_00, 'EUR')
to_currency = 'GBP'

bank = Money::Bank::Historical.instance

# exchange money with rates from December 10th 2016
bank.exchange_with_historical(from_money, to_currency, Date.new(2016, 12, 10))
# => #<Money fractional:8399 currency:GBP>

# can also pass a Time/DateTime object, it's converted into the respective UTC Date
bank.exchange_with_historical(from_money, to_currency, Time.utc(2016, 10, 2, 11, 0, 0))
# => #<Money fractional:8691 currency:GBP>
```

Or perform it directly on the `Money` object
```ruby
from_money.exchange_to_historical(to_currency, Date.new(2016, 12, 10))
# => #<Money fractional:8399 currency:GBP>
```

`Bank#exchange_with` and `Money#exchange_to` can still be used. In this case, recent rates are needed, so yesterday's closing rates are used for the calculation
```ruby
bank.exchange_with(from_money, to_currency)

# set the default bank and create a new Money object that will use it
Money.default_bank = Money::Bank::Historical.instance
from_money = Money.new(100_00, 'EUR')
from_money.exchange_to(to_currency)
```

### Adding and retrieving rates

Adding rates will not be needed in most cases as the rates are fetched from OER.
However, `#add_rate` and `#get_rate` were implemented in order to conform to the Bank API.
An extra `#add_rates` method was implemented for setting rates in bulk.

`#add_rate` and `add_rates` can prove quite when **testing**,
as you can't afford HTTP requests there.
Only thing you need to do is initialize the bank, and add some default rates
before your tests run.


#### Get single rate

`#get_rate` accepts both ISO strings and Money::Currency objects
```ruby
bank.get_rate(Money::Currency.new('GBP'), 'CAD', Date.new(2016, 10, 1))
# => #<BigDecimal:7fd39fd2cb78,'0.1703941289 451827243E1',27(45)>
```

Getting without a datetime will return yesterday's closing rate, e.g. `bank.get_rate('CAD', 'CNY')`.

#### Add single rate

`#add_rate` adds a single rate to the Redis cache. It accepts both ISO strings and `Money::Currency` objects. Added rates should be relative to the base currency.
```ruby
date = Date.new(2016, 5, 18)

bank.add_rate('EUR', 'USD', 1.2, date)
bank.add_rate(Money::Currency.new('USD'), Money::Currency.new('GBP'), 0.8, date)

# 100 EUR = 100 * 1.2 USD = 100 * 1.2 * 0.8 GBP = 96 GBP
bank.exchange_with_historical(from_money, to_currency, date)
# => #<Money fractional:9600 currency:GBP>
```

Adding without a datetime will set the rate to yesterday's closing rate
```ruby
bank.add_rate('EUR', 'USD', 1.4)
bank.add_rate(Money::Currency.new('USD'), Money::Currency.new('GBP'), 0.6)

# 100 EUR = 100 * 1.4 USD = 100 * 1.4 * 0.6 GBP = 84 GBP
bank.exchange_with(from_money, to_currency)
# => #<Money fractional:8400 currency:GBP>
```

Trying to add a rate that is not relative to the base currency will fail.
This is because all cached rates are relative to the base currency.
```ruby
bank.add_rate('EUR', 'GBP', 0.96, date)
# ArgumentError: `from_currency` (EUR) or `to_currency` (GBP) should match the base currency USD
```

#### Add rates in bulk

`#add_rates` can be used to add multiple historical rates (relative to the base currency) to the Redis cache.
```ruby
rates = {
  'EUR' => {
    '2015-09-10' => 0.11, # 1 USD = 0.11 EUR
    '2015-09-11' => 0.22
  },
  'GBP' => {
    '2015-09-10' => 0.44, # 1 USD = 0.44 GBP
    '2015-09-11' => 0.55
  }
}
bank.add_rates(rates)

# 100 EUR = 100 / 0.11 USD = 100 / 0.11 * 0.44 GBP = 400 GBP
bank.exchange_with_historical(from_money, to_currency, Date.new(2015, 9, 10))
# => #<Money fractional:40000 currency:GBP>
```


CHANGELOG.md
# Changelog

## 0.1.4
- Support `redis` gem versions '>=3.3', '~> 4.0'

## 0.1.3
- Support `redis` gem versions '>=3.3', '< 4.1' (#3)

## 0.1.0
- Added basic functionality and documentation.

## 0.1.1
- Added support for FREE and DEVELOPER OpenExchangeRates accounts. Downside when using these accounts is that OER rates are fetched for a single day at a time (as opposed to a whole month for more advanced plans).
- Developers have to specify `oer_account_type` during configuration.

## 0.1.2
- Updated `rubocop` gem needed for development due to low-severity security issue with version 0.46


spec/spec_helper.rb
#
# Copyright 2017 Skyscanner Limited.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# frozen_string_literal: true

ENV['RACK_ENV'] ||= 'test'
ENV['REDIS_PORT'] ||= '6390'

require 'webmock/rspec'
require 'faker'
require 'timecop'
require 'byebug'
require 'money/bank/historical'

RSpec.configure do |config|
end


spec/redis-server.sh
#!/bin/bash

# start redis server without persistence on $REDIS_PORT (if not set, start on 6390)
redis-server --save "" --appendonly no --port ${REDIS_PORT:-6390}


spec/bank/historical_spec.rb
#
# Copyright 2017 Skyscanner Limited.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# frozen_string_literal: true

require 'spec_helper'

class Money
  module Bank
    describe Historical do
      let(:base_currency) { Currency.new('EUR') }
      let(:redis_url) { "redis://localhost:#{ENV['REDIS_PORT']}" }
      let(:redis) { Redis.new(port: ENV['REDIS_PORT']) }
      let(:redis_namespace) { 'currency_test' }
      let(:bank) { Historical.instance }

      before do
        Historical.configure do |config|
          config.base_currency = base_currency
          config.redis_url = redis_url
          config.redis_namespace = redis_namespace
          config.oer_app_id = SecureRandom.hex
          config.timeout = 20
        end
      end

      after do
        keys = redis.keys("#{redis_namespace}*")
        redis.del(keys) unless keys.empty?
      end

      describe '#add_rates' do
        let(:usd_date_rate_hash) do
          {
            '2015-09-10' => 0.11111.to_d,
            '2015-09-11' => 0.22222.to_d,
            '2015-09-12' => 0.33333.to_d
          }
        end
        let(:currency_date_rate_hash) do
          {
            'USD' => usd_date_rate_hash,
            'GBP' => {
              '2015-09-10' => 0.44444.to_d,
              '2015-09-11' => 0.55555.to_d,
              '2015-09-12' => 0.66666.to_d
            },
            'VND' => {
              '2015-09-10' => 0.77777.to_d,
              '2015-09-11' => 0.88888.to_d,
              '2015-09-12' => 0.99999.to_d
            }
          }
        end
        # it's pointing to the same cache as the one in bank
        let(:new_store) do
          RatesStore::HistoricalRedis.new(base_currency, redis_url, redis_namespace)
        end

        subject { bank.add_rates(currency_date_rate_hash) }

        it 'sets the rates' do
          subject
          cached_rates = new_store.get_rates(Currency.new('USD'))

          expect(cached_rates).to eq usd_date_rate_hash
        end
      end

      describe '#add_rate' do
        let(:rate) { rand }

        # it's pointing to the same cache as the one in bank
        let(:new_store) do
          RatesStore::HistoricalRedis.new(base_currency, redis_url, redis_namespace)
        end
        let(:datetime) { Time.utc(2017, 1, 4, 13, 0, 0) }

        subject { bank.add_rate(from_currency, to_currency, rate, datetime) }

        context 'when base currency == from_currency' do
          let(:from_currency) { 'EUR' }
          let(:to_currency) { Money::Currency.new('USD') }

          it "sets date's rate" do
            subject
            cached_rates = new_store.get_rates(to_currency)

            cached_rate = cached_rates['2017-01-04']
            expect(cached_rate).to be_within(0.00000001).of(rate)
          end
        end

        context 'when base currency == to_currency' do
          let(:from_currency) { Money::Currency.new('USD') }
          let(:to_currency) { Money::Currency.new('EUR') }

          it "sets date's rate as the inverse" do
            subject
            cached_rates = new_store.get_rates(from_currency)

            cached_rate = cached_rates['2017-01-04']
            expect(cached_rate).to be_within(0.00000001).of(1 / rate)
          end
        end

        context 'when base currency does not match either of the passed currencies' do
          let(:from_currency) { Money::Currency.new('USD') }
          let(:to_currency) { 'GBP' }

          it 'fails with ArgumentError' do
            expect { subject }.to raise_error ArgumentError
          end
        end

        context 'when datetime is a Date' do
          let(:datetime) { Faker::Date.between(Date.today - 100, Date.today + 100) }
          let(:from_currency) { 'EUR' }
          let(:to_currency) { Money::Currency.new('USD') }

          it "sets yesterday's rate" do
            subject
            cached_rates = new_store.get_rates(to_currency)

            cached_rate = cached_rates[datetime.iso8601]
            expect(cached_rate).to be_within(0.00000001).of(rate)
          end
        end

        context 'when datetime is not passed' do
          let(:from_currency) { 'EUR' }
          let(:to_currency) { Money::Currency.new('USD') }
          let(:now) { Time.utc(2017, 12, 5, 13, 0, 0) }

          subject { bank.add_rate(from_currency, to_currency, rate) }

          before { Timecop.travel(now) }
          after { Timecop.return }

          it "sets yesterday's rate" do
            subject
            cached_rates = new_store.get_rates(to_currency)

            cached_rate = cached_rates['2017-12-04']
            expect(cached_rate).to be_within(0.00000001).of(rate)
          end
        end
      end

      describe '#get_rate' do
        let(:from_currency) { 'EUR' }
        let(:to_currency) { Money::Currency.new('USD') }
        let(:rate) { rand }

        context 'when datetime is passed' do
          subject { bank.get_rate(from_currency, to_currency, datetime) }

          before { bank.add_rate(from_currency, to_currency, rate, datetime) }

          context 'and it is a Time' do
            let(:datetime) { Time.utc(2016, 3, 12, 14, 0, 0) }

            it 'returns the correct rate' do
              expect(subject).to be_within(0.00000001).of(rate)
            end
          end

          context 'and it is a Date' do
            let(:datetime) { Faker::Date.between(Date.today - 100, Date.today + 100) }

            it 'returns the correct rate' do
              expect(subject).to be_within(0.00000001).of(rate)
            end
          end
        end

        context 'when datetime is not passed' do
          subject { bank.get_rate(from_currency, to_currency) }

          before { bank.add_rate(from_currency, to_currency, rate) }

          it "returns yesterday's rate" do
            expect(subject).to be_within(0.00000001).of(rate)
          end
        end
      end

      describe '#exchange_with_historical' do
        let(:from_currency) { Currency.wrap('USD') }
        let(:from_money) { Money.new(10_000, from_currency) }
        let(:to_currency) { Currency.wrap('GBP') }
        let(:datetime) { Faker::Time.between(Date.today - 300, Date.today - 2) }
        let(:date) { datetime.utc.to_date }
        let(:from_currency_base_rate) { 1.1250721881474421.to_d }
        let(:to_currency_base_rate) { 0.7346888078084116.to_d }
        let(:from_currency_base_rates) do
          {
            (date - 1).iso8601 => rand,
            date.iso8601       => from_currency_base_rate,
            (date + 1).iso8601 => rand
          }
        end
        let(:to_currency_base_rates) do
          {
            (date - 1).iso8601 => rand,
            date.iso8601       => to_currency_base_rate,
            (date + 1).iso8601 => rand
          }
        end
        let(:expected_result) { Money.new(6530, to_currency) }

        before do
          allow_any_instance_of(RatesStore::HistoricalRedis).to receive(:get_rates)
            .with(from_currency).and_return(from_currency_base_rates_store)
          allow_any_instance_of(RatesStore::HistoricalRedis).to receive(:get_rates)
            .with(to_currency).and_return(to_currency_base_rates_store)
          allow_any_instance_of(RatesProvider::OpenExchangeRates)
            .to receive(:fetch_rates).with(date).and_return(rates_provider)
        end

        subject { bank.exchange_with_historical(from_money, to_currency, datetime) }

        describe 'to_currency parameter' do
          let(:from_currency_base_rates_store) { from_currency_base_rates }
          let(:to_currency_base_rates_store) { to_currency_base_rates }
          let(:rates_provider) { nil }

          context 'when iso code string' do
            let(:to_currency) { 'GBP' }

            it { is_expected.to eq expected_result }
          end
        end

        describe 'datetime type' do
          let(:datetime) { Faker::Time.between(Date.today - 300, Date.today - 2) }
          let(:date) { datetime.utc.to_date }
          let(:from_currency_base_rates_store) { from_currency_base_rates }
          let(:to_currency_base_rates_store) { to_currency_base_rates }
          let(:rates_provider) { nil }

          it 'returns same result when passing a date or time on that date' do
            time_result = bank.exchange_with_historical(from_money, to_currency, datetime)
            date_result = bank.exchange_with_historical(from_money, to_currency, date)

            expect(time_result).to eq date_result
          end
        end

        describe 'selecting data source' do
          context 'when both rates exist in Redis' do
            let(:from_currency_base_rates_store) { from_currency_base_rates }
            let(:to_currency_base_rates_store) { to_currency_base_rates }
            let(:rates_provider) { nil }

            it { is_expected.to eq expected_result }
          end

          context "when from_currency rate doesn't exist in Redis" do
            let(:from_currency_base_rates_store) { nil }
            let(:to_currency_base_rates_store) { to_currency_base_rates }
            let(:rates_provider) { { from_currency.iso_code => from_currency_base_rates } }

            it { is_expected.to eq expected_result }
          end

          context 'when from_currency rate exists in Redis for other dates' do
            let(:from_currency_base_rates_other_dates) do
              {
                (date - 3).iso8601 => rand,
                (date - 2).iso8601 => rand,
                (date - 1).iso8601 => rand
              }
            end
            let(:from_currency_base_rates_store) { from_currency_base_rates_other_dates }
            let(:to_currency_base_rates_store) { to_currency_base_rates }
            let(:rates_provider) { { from_currency.iso_code => from_currency_base_rates } }

            it { is_expected.to eq expected_result }
          end

          context "when to_currency rate doesn't exist in Redis" do
            let(:from_currency_base_rates_store) { from_currency_base_rates }
            let(:to_currency_base_rates_store) { nil }
            let(:rates_provider) { { to_currency.iso_code => to_currency_base_rates } }

            it { is_expected.to eq expected_result }
          end

          context 'when to_currency rate exists in Redis for other dates' do
            let(:to_currency_base_rates_other_dates) do
              {
                (date + 1).iso8601 => rand,
                (date + 2).iso8601 => rand,
                (date + 3).iso8601 => rand
              }
            end
            let(:from_currency_base_rates_store) { from_currency_base_rates }
            let(:to_currency_base_rates_store) { to_currency_base_rates_other_dates }
            let(:rates_provider) { { to_currency.iso_code => to_currency_base_rates } }

            it { is_expected.to eq expected_result }
          end

          context 'when neither of the rates exists in Redis' do
            let(:from_currency_base_rates_store) { nil }
            let(:to_currency_base_rates_store) { nil }
            let(:rates_provider) do
              {
                from_currency.iso_code => from_currency_base_rates,
                to_currency.iso_code   => to_currency_base_rates
              }
            end

            it { is_expected.to eq expected_result }
          end

          context 'when from_currency == to_currency' do
            let(:to_currency) { from_money.currency }
            let(:from_currency_base_rates_store) { nil }
            let(:to_currency_base_rates_store) { nil }
            let(:rates_provider) { nil }
            let(:expected_result) { from_money }

            it { is_expected.to eq expected_result }
          end

          context 'when from_currency == base_currency' do
            let(:from_money) { Money.new(10_000, base_currency) }
            let(:from_currency_base_rates_store) { nil }
            let(:to_currency_base_rates_store) { to_currency_base_rates }
            let(:rates_provider) { nil }
            let(:expected_result) { Money.new(7347, to_currency) }

            it { is_expected.to eq expected_result }
          end

          context 'when to_currency == base_currency' do
            let(:to_currency) { base_currency }
            let(:from_currency_base_rates_store) { from_currency_base_rates }
            let(:to_currency_base_rates_store) { nil }
            let(:rates_provider) { nil }
            let(:expected_result) { Money.new(8888, to_currency) }

            it { is_expected.to eq expected_result }
          end
        end

        # taken from real rates from XE.com
        describe 'money conversion' do
          let(:from_currency_base_rates_store) { from_currency_base_rates }
          let(:to_currency_base_rates_store) { to_currency_base_rates }
          let(:rates_provider) { nil }

          context 'for rates example 1' do
            let(:from_currency) { Currency.wrap('USD') }
            let(:from_money) { Money.new(500_00, from_currency) }
            let(:to_currency) { Currency.wrap('GBP') }
            let(:from_currency_base_rate) { 1.13597.to_d }
            let(:to_currency_base_rate) { 0.735500.to_d }
            let(:expected_result) { Money.new(323_73, to_currency) }

            it { is_expected.to eq expected_result }
          end

          context 'for rates example 2' do
            let(:from_currency) { Currency.wrap('INR') }
            let(:from_money) { Money.new(6_516_200, from_currency) }
            let(:to_currency) { Currency.wrap('CAD') }
            let(:from_currency_base_rate) { 73.5602.to_d }
            let(:to_currency_base_rate) { 1.46700.to_d }
            let(:expected_result) { Money.new(1_299_52, to_currency) }

            it { is_expected.to eq expected_result }
          end

          # VND has no decimal places
          context 'for rates example 3' do
            let(:from_currency) { Currency.wrap('SGD') }
            let(:from_money) { Money.new(345_67, from_currency) }
            let(:to_currency) { Currency.wrap('VND') }
            let(:from_currency_base_rate) { 1.57222.to_d }
            let(:to_currency_base_rate) { 25_160.75.to_d }
            let(:expected_result) { Money.new(5_531_870, to_currency) }

            it { is_expected.to eq expected_result }
          end

          # KWD has 3 decimal places
          context 'for rates example 4' do
            let(:from_currency) { Currency.wrap('CNY') }
            let(:from_money) { Money.new(987_654, from_currency) }
            let(:to_currency) { Currency.wrap('KWD') }
            let(:from_currency_base_rate) { 7.21517.to_d }
            let(:to_currency_base_rate) { 0.342725.to_d }
            let(:expected_result) { Money.new(469_142, to_currency) }

            it { is_expected.to eq expected_result }
          end
        end

        context 'when OER client fails with ArgumentError' do
          let(:datetime) { Faker::Time.between(Date.new(1990, 1, 1), Date.new(1998, 12, 31)) }
          let(:from_currency_base_rates_store) { nil }
          let(:to_currency_base_rates_store) { nil }
          let(:rates_provider) { nil }

          before do
            # unstub and let it blow up
            allow_any_instance_of(RatesProvider::OpenExchangeRates)
              .to receive(:fetch_rates).and_call_original
          end

          it 'fails' do
            expect { subject }.to raise_error(ArgumentError)
          end
        end
      end

      describe '#exchange_with' do
        let(:from_currency) { Currency.wrap('USD') }
        let(:from_money) { Money.new(10_000, from_currency) }
        let(:to_currency) { Currency.wrap('GBP') }
        let(:utc_date) { Time.now.utc.to_date }
        let(:from_currency_base_rate) { 1.1250721881474421.to_d }
        let(:to_currency_base_rate) { 0.7346888078084116.to_d }
        let(:from_currency_base_rates_store) do
          {
            (utc_date - 3).iso8601 => rand,
            (utc_date - 2).iso8601 => rand,
            (utc_date - 1).iso8601 => from_currency_base_rate
          }
        end
        let(:to_currency_base_rates_store) do
          {
            (utc_date - 3).iso8601 => rand,
            (utc_date - 2).iso8601 => rand,
            (utc_date - 1).iso8601 => to_currency_base_rate
          }
        end
        let(:expected_result) { Money.new(6530, to_currency) }

        before do
          allow_any_instance_of(Money::RatesStore::HistoricalRedis).to receive(:get_rates)
            .with(from_currency).and_return(from_currency_base_rates_store)
          allow_any_instance_of(Money::RatesStore::HistoricalRedis).to receive(:get_rates)
            .with(to_currency).and_return(to_currency_base_rates_store)
        end

        subject { bank.exchange_with(from_money, to_currency) }

        it "selects yesterday's rates" do
          expect(subject).to eq expected_result
        end
      end
    end
  end

  describe Money do
    describe '#exchange_with_historical' do
      let(:base_currency) { Currency.new('EUR') }
      let(:redis_url) { "redis://localhost:#{ENV['REDIS_PORT']}" }
      let(:redis_namespace) { 'currency_test' }
      let(:rates) do
        {
          'USD' => { '2015-09-10' => 0.11 },
          'GBP' => { '2015-09-10' => 0.44 }
        }
      end
      let(:money) { Money.new(100_00, 'USD') }
      let(:to_currency) { Money::Currency.new('GBP') }
      let(:datetime) { Date.new(2015, 9, 10) }

      before do
        Bank::Historical.configure do |config|
          config.base_currency = base_currency
          config.redis_url = redis_url
          config.redis_namespace = redis_namespace
        end

        Bank::Historical.instance.add_rates(rates)
      end

      subject { money.exchange_to_historical(to_currency, datetime) }

      # on Sept 10th, 100 EUR = 100 / 0.11 USD = 100 / 0.11 * 0.44 GBP = 400 GBP
      it { is_expected.to eq Money.new(400_00, 'GBP') }
    end
  end
end


examples/add_get_rates.rb
#
# Copyright 2017 Skyscanner Limited.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# frozen_string_literal: true

require 'money/bank/historical'

redis_url = 'redis://localhost:6379'
namespace = 'currency_example'

Money::Bank::Historical.configure do |config|
  # (required) your OpenExchangeRates App ID
  config.oer_app_id = 'XXXXXXXXXXXXXXXXXXXXX'

  # (optional) currency relative to which all the rates are stored (default: EUR)
  config.base_currency = Money::Currency.new('USD')

  # (optional) the URL of the Redis server (default: 'redis://localhost:6379')
  config.redis_url = redis_url

  # (optional) Redis namespace to prefix all keys (default: 'currency')
  config.redis_namespace = namespace

  # (optional) set a timeout for the OER calls (default: 15 seconds)
  config.timeout = 20
end

bank = Money::Bank::Historical.instance

from_money = Money.new(100_00, 'EUR')
to_currency = 'GBP'

############ Get single rate #################

# It accepts both ISO strings and Money::Currency objects.
bank.get_rate(Money::Currency.new('GBP'), 'CAD', Date.new(2016, 10, 1))
# => #<BigDecimal:7fd39fd2cb78,'0.1703941289 451827243E1',27(45)>

# Getting without a datetime will return yesterday's closing rate
bank.get_rate('CAD', 'CNY')

############ Add single rate #################

# It accepts both ISO strings and Money::Currency objects.
# Added rates should be relative to the base currency
date = Date.new(2016, 5, 18)

bank.add_rate('EUR', 'USD', 1.2, date)
bank.add_rate(Money::Currency.new('USD'), Money::Currency.new('GBP'), 0.8, date)

# 100 EUR = 100 * 1.2 USD = 100 * 1.2 * 0.8 GBP = 96 GBP
bank.exchange_with_historical(from_money, to_currency, date)
# => #<Money fractional:9600 currency:GBP>

# Adding without a datetime will set the rate to yesterday's closing rate
bank.add_rate('EUR', 'USD', 1.4)
bank.add_rate(Money::Currency.new('USD'), Money::Currency.new('GBP'), 0.6)

# 100 EUR = 100 * 1.4 USD = 100 * 1.4 * 0.6 GBP = 84 GBP
bank.exchange_with(from_money, to_currency)
# => #<Money fractional:8400 currency:GBP>

# trying to add a rate that is not relative to the base currency will fail
bank.add_rate('EUR', 'GBP', 0.96, date)
# ArgumentError: `from_currency` (EUR) or `to_currency` (GBP) should match the base currency USD

############ Add rates in bulk #################

# add historical exchange rates (relative to the base currency) in bulk
rates = {
  'EUR' => {
    '2015-09-10' => 0.11, # 1 USD = 0.11 EUR
    '2015-09-11' => 0.22,
    '2015-09-12' => 0.33
  },
  'GBP' => {
    '2015-09-10' => 0.44, # 1 USD = 0.44 GBP
    '2015-09-11' => 0.55,
    '2015-09-12' => 0.66
  },
  'VND' => {
    '2015-09-10' => 0.77, # 1 USD = 0.77 VND
    '2015-09-11' => 0.88,
    '2015-09-12' => 0.99
  }
}
bank.add_rates(rates)

# 100 EUR = 100 / 0.11 USD = 100 / 0.11 * 0.44 GBP = 400 GBP
bank.exchange_with_historical(from_money, to_currency, Date.new(2015, 9, 10))
# => #<Money fractional:40000 currency:GBP>

########## Clean up Redis keys used here #########

redis = Redis.new(url: redis_url)
keys = redis.keys("#{namespace}*")
redis.del(keys)
